File: genprime.rb
#!/usr/bin/ruby

$primes = Array.new

def isprime(x)
	if x < 2
		return 0
	end
	if x == 2
		return 1
	end
	if x % 2 == 0
		return 0
	end
	if x < 9
		return 1
	end
	if (x + 1) % 6 != 0
		if (x - 1) % 6 != 0
			return 0
		end
	end
	lim=Math.sqrt(x).to_i+1
	y = 1
	$primes.each do |v|
		y = v
		if y >= lim
			return 1
		end
		if x % y == 0
			return 0
		end
	end
	y += 2
	while y <= lim
		if x % y == 0
			return 0
		end
		y += 2
	end
	return 1
end

def genprime(max)
	count = 0
	current = 1
	while count<max
		if isprime(current) != 0
			if $primes.length < 1 or $primes[-1] < current
				$primes << current
			end
			count += 1
		end
		current += 1
	end
	return current - 1
end

start = ARGV[0].to_i
stop = ARGV[1].to_i + 1
i = start
if start == 0
	Process.exit
end
until i >= stop
	starttime = Time.now
	last = genprime(i)
	endtime = Time.now
	duration = endtime - starttime
	printf "Found %8d primes in %10.5f seconds (last was %10d)\n", i, duration, last
	i += start
end

File: genprime.py
#!/usr/bin/env python

import time
import sys
from math import sqrt

def isprime(x):
	if x < 2:
		return False
	if x == 2:
		return True
	if x % 2 == 0:
		return False
	if x < 9:
		return True
	if (x + 1) % 6 != 0:
		if (x - 1) % 6 != 0:
			return False
	lim=sqrt(x)+1
	for y in range(3, int(lim), 2):
		if x % y == 0:
			return False
	return True

def genprime(max):
	count = 0
	current = 1
	while count<max:
		if isprime(current):
			count = count + 1
		current = current + 1
	return current - 1

def main():
	ind=1
	if sys.argv[1] == '-p':
		ind = ind + 1
		import psyco
		psyco.full()
	try:
		start=int(sys.argv[ind])
		stop=int(sys.argv[ind + 1])+1
	except IndexError:
		quit()
	for x in range(start, stop, start):
		begin = time.clock()
		last = genprime(x)
		end = time.clock()
		duration = end - begin
		print('Found %(count)8d primes in %(time)10.5f seconds (last was %(last)10d)' % {'count': x, 'time': duration, 'last': last})

if __name__ == "__main__":
	main()

